lib/luaxtable.c

```diff
@@ -32,48 +14,165 @@
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret)
 {
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		goto err;
+	}
 	return 0;
+err:
+	return -1;
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static inline lunatik_object_t *luaxtable_getskb(lua_State *L, luaxtable_t *xtable)
 {
+	if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA)
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+static int luaxtable_pushparams(lua_State *L, const struct xt_action_param *par, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
-	return NF_ACCEPT;
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+	luadata_reset(data, skb->data, skb_headlen(skb), opt);
+
+	lua_newtable(L);
+	lua_pushboolean(L, par->hotdrop);
+	lua_setfield(L, -2, "hotdrop");
+	lua_pushinteger(L, par->thoff);
+	lua_setfield(L, -2, "thoff");
+	lua_pushinteger(L, par->fragoff);
+	lua_setfield(L, -2, "fragoff");
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 10, 0))
+	lua_pushinteger(L, xt_hooknum(par));
+#else
+	lua_pushinteger(L, par->hooknum);
+#endif
+	lua_setfield(L, -2, "hooknum");
+	lua_pushvalue(L, -1); /* param table */
+	lua_insert(L, lua_gettop(L) - 2); /* stack: (...), param, skb, param */
+	return 0;
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
+#define luaxtable_call(L, xtable, par, skb, op, opt) \
+	luaxtable_pushparams(L, par, xtable, skb, LUADATA_OPT_READONLY) == -1 || \
+	luaxtable_invoke(L, xtable, op, 2, 1) == -1
+
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par, int fallback)
 {
-	return 0;
+	if (luaxtable_call(L, xtable, par, skb, "match", LUADATA_OPT_READONLY))
+		return fallback;
+
+	int ret = lua_toboolean(L, -1);
+	lua_getfield(L, -2, "hotdrop");
+	par->hotdrop = lua_toboolean(L, -1);
+	return ret;
 }
 
-static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+static int luaxtable_dotarget(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const struct xt_action_param *par, int fallback)
 {
+	if (luaxtable_call(L, xtable, par, skb, "match", LUADATA_OPT_READONLY))
```

Comment:
I think `!= 0` or `< 0` would be a bit clearer..
```suggestion
	if (luaxtable_call(L, xtable, par, skb, "match", LUADATA_OPT_READONLY) != 0)
```